trestle/core/markdown/control_markdown_node.py

```diff
@@ -0,0 +1,464 @@
+# -*- mode:python; coding:utf-8 -*-
+
+# Copyright (c) 2023 IBM Corp. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A control markdown node."""
+from __future__ import annotations
+
+import logging
+import string
+from typing import List, Optional, Tuple
+
+from trestle.common import const
+from trestle.common.err import TrestleError
+from trestle.common.list_utils import none_if_empty
+from trestle.core.control_interface import ControlInterface
+from trestle.core.markdown.base_markdown_node import BaseMarkdownNode, BaseSectionContent
+from trestle.oscal import common
+
+logger = logging.getLogger(__name__)
+
+
+class TreeContext:
+    """A shared singleton tree context."""
+
+    def __init__(self):
+        """Initialize tree context."""
+        self.control_id = ''
+        self.control_group = ''
+        self.control_title = ''
+
+    def reset(self):
+        """Reset global control tree context."""
+        self.control_id = ''
+        self.control_group = ''
+        self.control_title = ''
+
+
+tree_context = TreeContext()
+
+
+class ControlSectionContent(BaseSectionContent):
+    """A content of the node."""
+
+    def __init__(self):
+        """Initialize section content."""
+        super(ControlSectionContent, self).__init__()
+        self.statement_part = None
+        self.objective_part = None
+        self.guidance_part = None
+        self.other_parts = []
+
+    def union(self, node: ControlMarkdownNode) -> None:
+        """Unites contents together."""
+        super().union(node)
+        pass
+
+
+class ControlMarkdownNode(BaseMarkdownNode):
+    """Markdown will be read to the tree."""
+
+    def __init__(self, key: str, content: ControlSectionContent, starting_line: int):
+        """Initialize markdown node."""
+        super(ControlMarkdownNode, self).__init__(key, content, starting_line)
+        self.content: ControlSectionContent = content
+
+    def _build_tree(
+        self,
```

Comment:
This has very high cognitive complexity and if it can be reduced in a helpful way that would be good.  But if the overall flow is clear it may be ok.

Reply:
Will try to address this